File: tests/shared/test_helpers.js
/**
 * Shared test helpers for JavaScript/React tests
 */

import { render } from '@testing-library/react';
import { jest } from '@jest/globals';

/**
 * Create mock axios instance for testing
 */
export const createMockAxios = () => {
  return {
    get: jest.fn(),
    post: jest.fn(),
    put: jest.fn(),
    delete: jest.fn(),
    patch: jest.fn(),
    defaults: {
      headers: {
        common: {}
      }
    }
  };
};

/**
 * Mock question data for testing
 */
export const mockQuestions = [
  {
    id: 1,
    text: 'What is the capital of France?',
    options: [
      { id: 'a', text: 'London' },
      { id: 'b', text: 'Berlin' },
      { id: 'c', text: 'Paris' },
      { id: 'd', text: 'Madrid' }
    ],
    correct_answer: 'c',
    category: 'geography'
  },
  {
    id: 2,
    text: 'What is 2 + 2?',
    options: [
      { id: 'a', text: '3' },
      { id: 'b', text: '4' },
      { id: 'c', text: '5' },
      { id: 'd', text: '6' }
    ],
    correct_answer: 'b',
    category: 'math'
  },
  {
    id: 3,
    text: 'Which programming language is used for web development?',
    options: [
      { id: 'a', text: 'Python' },
      { id: 'b', text: 'JavaScript' },
      { id: 'c', text: 'Java' },
      { id: 'd', text: 'C++' }
    ],
    correct_answer: 'b',
    category: 'programming'
  }
];

/**
 * Mock AI-generated questions
 */
export const mockAIQuestions = [
  {
    text: 'What is machine learning?',
    options: [
      { id: 'a', text: 'A type of computer hardware' },
      { id: 'b', text: 'A subset of artificial intelligence' },
      { id: 'c', text: 'A programming language' },
      { id: 'd', text: 'A database system' }
    ],
    correct_answer: 'b',
    category: 'technology'
  }
];

/**
 * Mock quiz results
 */
export const mockQuizResults = {
  totalQuestions: 3,
  correctAnswers: 2,
  scorePercentage: 66.67,
  answers: [
    { questionId: 1, selectedAnswer: 'c', isCorrect: true },
    { questionId: 2, selectedAnswer: 'a', isCorrect: false },
    { questionId: 3, selectedAnswer: 'b', isCorrect: true }
  ]
};

/**
 * Mock configuration data
 */
export const mockConfig = {
  llmProvider: 'openai',
  openaiModel: 'gpt-3.5-turbo',
  ollamaModel: 'llama3.2',
  ollamaHost: 'http://localhost:11434'
};

/**
 * Mock logging configuration
 */
export const mockLoggingConfig = {
  level: 'INFO',
  format: '%(asctime)s - %(levelname)s - %(message)s',
  handlers: {
    file: {
      filename: 'app.log',
      maxBytes: 1024000,
      backupCount: 3
    },
    console: {
      stream: 'stdout'
    }
  }
};

/**
 * Create a test question with default values
 */
export const createTestQuestion = ({
  id = 1,
  text = 'Test question?',
  category = 'test',
  correctAnswer = 'a',
  options = [
    { id: 'a', text: 'Option A' },
    { id: 'b', text: 'Option B' },
    { id: 'c', text: 'Option C' },
    { id: 'd', text: 'Option D' }
  ]
} = {}) => {
  return {
    id,
    text,
    category,
    correct_answer: correctAnswer,
    options
  };
};

/**
 * Create mock API responses
 */
export const createMockApiResponse = (data, status = 200) => {
  return {
    data,
    status,
    statusText: status === 200 ? 'OK' : 'Error',
    headers: {},
    config: {}
  };
};

/**
 * Create mock error response
 */
export const createMockErrorResponse = (message = 'API Error', status = 500) => {
  const error = new Error(message);
  error.response = {
    status,
    statusText: 'Error',
    data: { error: message }
  };
  return Promise.reject(error);
};

/**
 * Validate question structure
 */
export const validateQuestionStructure = (question) => {
  const requiredFields = ['text', 'options', 'correct_answer'];
  
  for (const field of requiredFields) {
    if (!(field in question)) {
      return false;
    }
  }
  
  if (!Array.isArray(question.options) || question.options.length === 0) {
    return false;
  }
  
  for (const option of question.options) {
    if (!('id' in option) || !('text' in option)) {
      return false;
    }
  }
  
  // Check that correct_answer references a valid option
  const optionIds = question.options.map(opt => opt.id);
  if (!optionIds.includes(question.correct_answer)) {
    return false;
  }
  
  return true;
};

/**
 * Calculate quiz score
 */
export const calculateQuizScore = (answers) => {
  if (!answers || answers.length === 0) {
    return 0;
  }
  
  const correctCount = answers.filter(answer => answer.isCorrect).length;
  return (correctCount / answers.length) * 100;
};

/**
 * Wait for async operations in tests
 */
export const waitFor = (condition, timeout = 5000) => {
  return new Promise((resolve, reject) => {
    const startTime = Date.now();
    
    const check = () => {
      if (condition()) {
        resolve();
      } else if (Date.now() - startTime > timeout) {
        reject(new Error(`Timeout waiting for condition after ${timeout}ms`));
      } else {
        setTimeout(check, 100);
      }
    };
    
    check();
  });
};

/**
 * Mock local storage for testing
 */
export const createMockLocalStorage = () => {
  let store = {};
  
  return {
    getItem: jest.fn((key) => store[key] || null),
    setItem: jest.fn((key, value) => {
      store[key] = value.toString();
    }),
    removeItem: jest.fn((key) => {
      delete store[key];
    }),
    clear: jest.fn(() => {
      store = {};
    }),
    get length() {
      return Object.keys(store).length;
    },
    key: jest.fn((index) => Object.keys(store)[index] || null)
  };
};

/**
 * Mock session storage for testing
 */
export const createMockSessionStorage = () => {
  return createMockLocalStorage(); // Same interface
};

/**
 * Setup global mocks for tests
 */
export const setupGlobalMocks = () => {
  // Mock window.localStorage
  Object.defineProperty(window, 'localStorage', {
    value: createMockLocalStorage(),
    writable: true
  });
  
  // Mock window.sessionStorage
  Object.defineProperty(window, 'sessionStorage', {
    value: createMockSessionStorage(),
    writable: true
  });
  
  // Mock window.location
  delete window.location;
  window.location = {
    href: 'http://localhost:3000',
    origin: 'http://localhost:3000',
    protocol: 'http:',
    host: 'localhost:3000',
    hostname: 'localhost',
    port: '3000',
    pathname: '/',
    search: '',
    hash: '',
    assign: jest.fn(),
    replace: jest.fn(),
    reload: jest.fn()
  };
  
  // Mock console methods to reduce noise in tests
  global.console = {
    ...console,
    warn: jest.fn(),
    error: jest.fn(),
    log: jest.fn()
  };
};

/**
 * Cleanup after tests
 */
export const cleanupAfterTests = () => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
};

/**
 * Custom render function with common providers
 */
export const renderWithProviders = (ui, options = {}) => {
  // In a real app, you might wrap with providers like:
  // - React Router
  // - Redux/Context providers
  // - Theme providers
  // - Error boundaries
  
  const Wrapper = ({ children }) => {
    // Add providers here as needed
    return children;
  };
  
  return render(ui, { wrapper: Wrapper, ...options });
};

/**
 * Common test categories
 */
export const testCategories = [
  'geography',
  'math', 
  'programming',
  'science',
  'technology',
  'history'
];

/**
 * API endpoints for testing
 */
export const apiEndpoints = {
  questions: '/api/questions',
  aiQuestions: '/api/questions/ai',
  categories: '/api/categories',
  loggingConfig: '/api/logging/config',
  llmConfig: '/api/llm/config',
  health: '/api/health'
};

// Export all helpers
export default {
  createMockAxios,
  mockQuestions,
  mockAIQuestions,
  mockQuizResults,
  mockConfig,
  mockLoggingConfig,
  createTestQuestion,
  createMockApiResponse,
  createMockErrorResponse,
  validateQuestionStructure,
  calculateQuizScore,
  waitFor,
  createMockLocalStorage,
  createMockSessionStorage,
  setupGlobalMocks,
  cleanupAfterTests,
  renderWithProviders,
  testCategories,
  apiEndpoints
};

File: tests/shared/test_helpers.py
"""Shared test helpers for Python tests."""

import os
import sys
import tempfile
import json
import sqlite3
from typing import Dict, Any, List, Optional
from contextlib import contextmanager

# Add backend directory to path for imports
BACKEND_DIR = os.path.join(os.path.dirname(__file__), '..', '..', 'backend')
sys.path.insert(0, BACKEND_DIR)

class DatabaseTestHelper:
    """Helper class for database-related testing."""
    
    def __init__(self, db_path: str = ":memory:"):
        self.db_path = db_path
        self.connection = None
    
    def setup_test_database(self):
        """Set up a test database with sample data."""
        self.connection = sqlite3.connect(self.db_path)
        cursor = self.connection.cursor()
        
        # Create questions table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS questions (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                text TEXT NOT NULL,
                category TEXT NOT NULL,
                correct_answer TEXT NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        ''')
        
        # Create options table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS options (
                id TEXT NOT NULL,
                question_id INTEGER NOT NULL,
                text TEXT NOT NULL,
                FOREIGN KEY (question_id) REFERENCES questions (id)
            )
        ''')
        
        # Insert sample data
        sample_questions = [
            (1, "What is the capital of France?", "geography", "c"),
            (2, "What is 2 + 2?", "math", "b"),
            (3, "Which language is used for web development?", "programming", "b")
        ]
        
        cursor.executemany(
            "INSERT OR REPLACE INTO questions (id, text, category, correct_answer) VALUES (?, ?, ?, ?)",
            sample_questions
        )
        
        sample_options = [
            ("a", 1, "London"),
            ("b", 1, "Berlin"),
            ("c", 1, "Paris"),
            ("d", 1, "Madrid"),
            ("a", 2, "3"),
            ("b", 2, "4"),
            ("c", 2, "5"),
            ("d", 2, "6"),
            ("a", 3, "Python"),
            ("b", 3, "JavaScript"),
            ("c", 3, "Java"),
            ("d", 3, "C++")
        ]
        
        cursor.executemany(
            "INSERT OR REPLACE INTO options (id, question_id, text) VALUES (?, ?, ?)",
            sample_options
        )
        
        self.connection.commit()
        return self.connection
    
    def cleanup(self):
        """Clean up database connection."""
        if self.connection:
            self.connection.close()
    
    def get_question_count(self) -> int:
        """Get total number of questions in database."""
        cursor = self.connection.cursor()
        cursor.execute("SELECT COUNT(*) FROM questions")
        return cursor.fetchone()[0]
    
    def get_questions_by_category(self, category: str) -> List[Dict]:
        """Get questions for a specific category."""
        cursor = self.connection.cursor()
        cursor.execute(
            "SELECT id, text, category, correct_answer FROM questions WHERE category = ?",
            (category,)
        )
        questions = []
        for row in cursor.fetchall():
            question = {
                "id": row[0],
                "text": row[1],
                "category": row[2],
                "correct_answer": row[3]
            }
            
            # Get options for this question
            cursor.execute(
                "SELECT id, text FROM options WHERE question_id = ?",
                (question["id"],)
            )
            question["options"] = [
                {"id": opt_row[0], "text": opt_row[1]}
                for opt_row in cursor.fetchall()
            ]
            questions.append(question)
        
        return questions

class ConfigTestHelper:
    """Helper class for configuration testing."""
    
    @staticmethod
    def create_temp_config(config_data: Dict[str, Any] = None) -> str:
        """Create a temporary configuration file."""
        if config_data is None:
            config_data = {
                "llm_provider": "openai",
                "openai_model": "gpt-3.5-turbo",
                "openai_api_key": "test-key",
                "ollama_model": "llama3.2",
                "ollama_host": "http://localhost:11434"
            }
        
        temp_file = tempfile.NamedTemporaryFile(mode='w', suffix='.json', delete=False)
        json.dump(config_data, temp_file, indent=2)
        temp_file.close()
        return temp_file.name
    
    @staticmethod
    def cleanup_temp_config(config_path: str):
        """Remove temporary configuration file."""
        try:
            os.remove(config_path)
        except FileNotFoundError:
            pass

class APITestHelper:
    """Helper class for API testing."""
    
    @staticmethod
    def create_mock_response(data: Any, status_code: int = 200):
        """Create a mock HTTP response."""
        class MockResponse:
            def __init__(self, data, status_code):
                self.data = data
                self.status_code = status_code
            
            def json(self):
                return self.data
            
            def raise_for_status(self):
                if self.status_code >= 400:
                    raise Exception(f"HTTP {self.status_code} Error")
        
        return MockResponse(data, status_code)
    
    @staticmethod
    def validate_question_structure(question: Dict) -> bool:
        """Validate that a question has the required structure."""
        required_fields = ["text", "options", "correct_answer"]
        
        for field in required_fields:
            if field not in question:
                return False
        
        if not isinstance(question["options"], list) or len(question["options"]) == 0:
            return False
        
        for option in question["options"]:
            if "id" not in option or "text" not in option:
                return False
        
        # Check that correct_answer references a valid option
        option_ids = [opt["id"] for opt in question["options"]]
        if question["correct_answer"] not in option_ids:
            return False
        
        return True

class LoggingTestHelper:
    """Helper class for logging-related testing."""
    
    @staticmethod
    def create_temp_log_file() -> str:
        """Create a temporary log file for testing."""
        temp_file = tempfile.NamedTemporaryFile(mode='w', suffix='.log', delete=False)
        temp_file.close()
        return temp_file.name
    
    @staticmethod
    def cleanup_temp_log_file(log_path: str):
        """Remove temporary log file."""
        try:
            os.remove(log_path)
        except FileNotFoundError:
            pass
    
    @staticmethod
    def get_log_entries(log_path: str) -> List[str]:
        """Read log entries from a log file."""
        try:
            with open(log_path, 'r') as f:
                return f.readlines()
        except FileNotFoundError:
            return []

@contextmanager
def temporary_environment(**env_vars):
    """Context manager for temporarily setting environment variables."""
    old_environ = dict(os.environ)
    os.environ.update(env_vars)
    try:
        yield
    finally:
        os.environ.clear()
        os.environ.update(old_environ)

def assert_quiz_score(answers: List[Dict], expected_score: float, tolerance: float = 0.01):
    """Assert that quiz score calculation is correct."""
    correct_count = sum(1 for answer in answers if answer.get("is_correct", False))
    total_count = len(answers)
    calculated_score = (correct_count / total_count) * 100 if total_count > 0 else 0
    
    assert abs(calculated_score - expected_score) < tolerance, \
        f"Expected score {expected_score}, got {calculated_score}"

def create_test_question(
    text: str = "Test question?",
    category: str = "test",
    correct_answer: str = "a",
    options: List[Dict] = None
) -> Dict:
    """Create a test question with default values."""
    if options is None:
        options = [
            {"id": "a", "text": "Option A"},
            {"id": "b", "text": "Option B"},
            {"id": "c", "text": "Option C"},
            {"id": "d", "text": "Option D"}
        ]
    
    return {
        "text": text,
        "category": category,
        "correct_answer": correct_answer,
        "options": options
    }

# Export commonly used helpers
__all__ = [
    'DatabaseTestHelper',
    'ConfigTestHelper', 
    'APITestHelper',
    'LoggingTestHelper',
    'temporary_environment',
    'assert_quiz_score',
    'create_test_question'
]